doc/usecases/record_elektra/UC_ansible-libelektra_merge_strategies.md

```diff
@@ -0,0 +1,18 @@
+# Use Case: Select merge strategy in ansible-libelektra
+
+## Summary
+
+- **Title:** Select merge-strategy in ansible-libelektra
+- **Scope:** ansible-libelektra
+- **Level:** User goal
+- **Actors:** User (usually sysadmin)
+- **Brief:** It should be possible to specify the merge strategy used, e.g. OURS, THEIRS, ABORT. ansible-libelektra should then use this strategy when merging.
+
+## Scenarios
+
+- **Precondition:** -
+- **Main success scenario:** If the user has specified a strategy, this strategy is applied on merge conflicts.
+- **Alternative scenario:** If no strategy is specified, a sane (and documented) default strategy should be used.
```

Comment:
This is vague. It would be better if you decide.

Reply:
It think we should use OURS as default. It's probably the intention in 99% of the use cases to set a new value when using Ansible.

Reply:
I agree!

Reply:
I'm not sure, using "ours" to resolve conflicts by default is a good idea. IMO the default should be to just abort. Otherwise things may already be broken before the user even notices that there was a merge.

Yes, most people will probably decide to use "ours", but it's not a safe default IMO. It's kind of as if the default for an SQL INSERT would be to replace a conflicting primary key instead of aborting.

Reply:
Yes, abort on conflicts is probably a better choice. Then people look at which changes were done by maintainers and manually select what is actually best.

Btw. we discarded the undo UC. It might help with unwanted merges but it is probably not worth the effort to keep previous recordings.

Reply:
I'm not even sure anymore if we need the merging for the Ansible Use Case. If I execute an Anisble playbook, my goal is that all the hosts I run them on are in the configured state. After all, an Ansible playbook is supposed to be a single source of truth.

Reply:
I mean I'll still implement this feature ... but for the Case Studies this shouldn't be needed.
Also, to be usable we need to supply the `base`. Otherwise, it will always come to conflicts (except keys that never change) if we use `ABORT` as default.

Reply:
As I envisioned it, Ansible runs would set a new BASE. Modifications of Ansible playbooks are OURS, modifications on the machine are THEIRS.

THEIRS can have two reasons:
1. manual interventions as supported by our recording approach
2. updates of software (i.e. new config files in /etc installed by packages)

Because of case 2. this UC is very much needed whenever you will use ansible-libelektra (regardless of if you also use record-libelektra). If 1. is done correctly (all changes exported and added to playbooks) there actually should be no conflict, as OURS would be THEIRS (both different from BASE). If someone forgot exporting and the playbook introduces incompatible changes, the strategy suggested from @kodebach, would loudly communicate that the sys-admins did something wrong, which looks like to be the proper way to do it. (Careful sys-admins would reset the machines to the previous BASE if they played around and nothing useful came out.)

Did this clarify the concerns?

Reply:
Yes this did clarify a few concerns. I totally forgot about the updates of software.

The only thing I still don't get is what you mean by 'Ansible runs would set a new BASE'. Where is this BASE stored? Within the playbook?

Reply:
It cannot be the playbook (otherwise you would not be able to reset to base locally) but where exactly is actually implementation detail. I thought that when `kdb record` gets reset, "/elektra/record" (or whatever we use as path for recording) is empty and with every changed key, the old values/metadata of that key get stored in "/elektra/record". So basically the merge from the current keys and (with higher priority) /elektra/record is the base. But let us discuss details later when you actually come to the implementation. For now hooks/global plugins have highest priority :rocket:

Reply:
This is actually a good idea - but remember: We don't track mounted configuration files. So if an update changes those directly we are currently out of luck.

I agree, let's get this merged and concentrate on the hooks.